Apps/Sandcastle/gallery/Drawing on Terrain.html

```diff
@@ -0,0 +1,185 @@
+function startup(Cesium) {
+    'use strict';
+//Sandcastle_Begin
+var viewer = new Cesium.Viewer('cesiumContainer',  {
+    selectionIndicator : false,
+    terrainProvider : Cesium.createWorldTerrain()
+});
+// Create a function to place a point given a world position.
+function createPoint(worldPosition) {
+    var point = new Cesium.Entity({
+        position : worldPosition,
+        point : {
+            color : Cesium.Color.WHITE,
+            pixelSize : 5,
+            heightReference: Cesium.HeightReference.CLAMP_TO_GROUND
+        }
+    });
+    viewer.entities.add(point);
+    return point;
+}
+// On left mouse click, we want to:
+// * Start a new shape if one doesn't already exist.
+// * Add a vertex to the active shape if one exists.
+var activeShapePoints = [];
+var activeShape;
+var floatingPoint;
+var drawingMode = 'line';
+var handler = new Cesium.ScreenSpaceEventHandler(viewer.canvas);
+handler.setInputAction(function(event) {
+    // We use `viewer.scene.pickPosition` here instead of `viewer.camera.pickEllipsoid` so that
+    // we get the correct point when mousing over terrain.
+    var earthPosition = viewer.scene.pickPosition(event.position);
+    // `earthPosition` will be undefined if our mouse is not over the globe.
+    if (Cesium.defined(earthPosition)) {
+        // If this is the first point, we need to initialize the shape.
+        if (activeShapePoints.length === 0) {
+            // This is the point that'll follow the mouse.
+            floatingPoint = createPoint(earthPosition);
+            // For the first point, we want to push an extra position into our list of vertices
+            // this will be a dynamic point that'll be replaced when the mouse moves.
+            activeShapePoints.push(earthPosition);
+            // We create a callback property for our positions since we want the shape to update
+            // in real time as we move the mouse. We make sure to pass `false` to the
+            // `CallbackProperty` constructor to mark it as not constant.
+            var dynamicPositions = new Cesium.CallbackProperty(function () {
+                // This is a work-around since polylines when `clampToGround` is true will
+                // crash if you try to draw a zero length line
+                if (activeShapePoints.length === 2) {
+                    var distance = Cesium.Cartesian3.distance(activeShapePoints[0], activeShapePoints[1]);
+                    if (distance === 0) {
+                        return [];
+                    }
+                }
+                return activeShapePoints;
+            }, false);
+            if (drawingMode === 'line') {
+                activeShape = viewer.entities.add({
+                    polyline : {
+                        positions : dynamicPositions,
+                        clampToGround : true,
+                        width : 3
+                    }
+                });
+            }
+            if (drawingMode === 'polygon') {
```

Comment:
This structure of two independent `if` statements seem odd to me, even though only one will be exected-- We should use one behavior or the other

Since we only have two drawing modes, do you think a boolean would be a better way than a string to keep track of which mode?

Reply:
I've moved this into a function as per @hpinkos 's suggestion, but I don't think we should make it just a boolean. If it were just something like `drawingLines`, then to me it's not obvious that when that's false then it draws polygons as opposed to something else. I think it makes more sense for it to be an explicit mode.

I did make it an `else if` since you are right that there's no scenario where you want both to run.

Reply:
OK, I see your point. In that case, maybe it's worth renaming the mode and label from "line" to "polyline" to match the entity type used?